### DISCUSSION TITLE
일반함수와 익명 함수

### DISCUSSION THREAD

**Discussion by baechanghyeon** [WantedPreonboardingFE5team/pre-onboarding-7th-2-1-5#30] [Reference]
### **일반함수와 익명함수의 차이점** 
1. 일반 함수는 전역적이며, 전체가 다 호이스팅 되므로 호출의 위치와 구현의 위치간에 연관관계가 없다. 또한 재사용될 기능에 주로 사용된다.
2. 익명 함수는 선언부만 호이스팅되며 호출의 위치와 구현의 위치간에 순서가 알맞아야 하고, 한번만 사용하는 기능에 사용된다.
위 말만 보면 익명 함수가 오히려 번거로운 것처럼 보일 수 있지만, 이는 메모리 관리에 효과적인 방안이 될 수 있다.

일반 함수는 자바스크립트를 초기에 읽어올 때 모두 호이스팅 된다고 하였다. 만약, 전체 자바스크립트 내에서 단 한번만 쓰이는 함수가 일반함수로 구현되어있을경우, 메모리 사용량이 성능에 중요한 영향을 미칠 수 있는 웹 애플리케이션에서, 이는 메모리 낭비라고 볼 수 있다.

단 한번만 사용되는 (재사용이 필요없는) 함수의 경우, 불필요한 시간동안 메모리를 차지하지않고 않도록 익명함수로 구현한다면, 정확히 해당 함수가 필요한 위치에서만 해당함수가 필요한 위치에서만 해당함수가 구현되고 사라지면서 메모리를 아낄 수 있게 된다.

### **프로젝트 적용 단계**

하나의 상태값만 변경할 시 `(onChange={(e) => setState(e.target.value)}` 이와 같이 작성을 해도 되지만, 이와같이 작성하는 경우에는 팀원들과의 컨벤션이 이루어져야 하며, 실무에서는 이와같이 작성하여 콜백하는 경우를 지양하기도 한다. 
반대로, 일반 함수의 경우에는 아래와 같이 표현할 시 이 이벤트가 어떤 로직을 가지고 있는지에 대해 명시적으로 파악이 가능하다는 이점을 가지고 있다.
```
const onChangeHandler = (e) => {
	setState( e.target.value);
}
```
### **중복 코드 줄이기**
위와 같은 상태 변경 이벤트가 반복되는 경우에는 CustomHook을 사용하여 코드의 중복을 줄인다. (Ex. 회원가입 폼과 같이 여러개의 Input이 존재하는 경우)

- App.js
```
const [value, changeValue ] = useInput(“”);

<input type=“text” value={value} onChange={changeValue} />

```

- useInput.js
```
const useInput = (initialValue) => {
	const [value, setValue] = useState(initialValue);
	const handleChange =  (event) => {
		   setValue(event.target.value)	
	}
        return {
	           value,
	           onChange: handleChange
	}
```

참고 URL : https://designcode.io/react-hooks-handbook-useinput-hook

위와 같이 커스텀 훅인 useInput()을 만들어 반복되는 코드를 줄인다. 

### 후기
- 코드를 작성하면서 내가 한번에 알아보는 부분도 중요하지만 다른사람이 내 코드를 볼 때 쉽게 파악하는 것도 중요하다. ( 가독성 )
- 코드의 중복을 줄이자 ! 

지난번 코드리뷰 당시 익명함수 부분에 대해 잘 몰라서 정리 해봤습니다!   참고용 정도로만 봐주세요!

**Comment by jaeyeong815** [comment]
custom Hook 내용 잘 읽고 갑니다~,~
참고 URL 들어가서 한번 읽어봤는데 올려주신 내용에 있는 App.js와 같이 value, onChange를 다 명시하는 방법이 있고
명시하지 않고 스프레드 문법으로 {...email} 요렇게 전달하는 방법도 있네요!

스프레드 문법으로 사용하면 코드의 길이가 줄어든다는 장점이 있지만 정확히 어떤 속성을 전달하는지 모호한 단점이 있을 것 같아요.
하지만 사용하면 또 다를 수 있으니 이방법 저방법 써보면서 팀원들과 좋은 방법을 찾아가면 좋겠네요 ~,~

좋은 글 공유 감사합니다
